### ISSUE TITLE
BUG: meson fails to properly detect numpy, pybind11 and (sort of) pythran when cross compiling

### ISSUE THREAD

**Issue by ahesford** [scipy/scipy#16783]
The new dependency detection in [`scipy/meson.build`](https://github.com/scipy/scipy/blob/main/scipy/meson.build#L40-L92) works suitably for native builds of SciPy, but does not correctly detect build-arch dependencies for cross compilation. The meson rules invoke the host Python interpreter, there called `py3`, to import the `numpy`, `pybind11` and `pythran` packages (all installed for the host arch) and deduce the proper paths for includes and, in the case of NumPy, infer a search path for the `npymath` and `npyrandom` libraries.

> NOTE: the Pythran search is incorrect but doesn't really cause a problem because the it is only used on the host to render files; the include path returned by the problematic search is never referenced by any subsequent meson build logic. The entire `incdir_pythran` block could be removed without adverse effects.

If the requisite dependencies are installed for the host arch (whether or not they are also installed for the build arch), the detection will return paths for the host arch, which might cause subtle problems in header files and fails outright when the linker attempts to link object files for the build arch with the native `npymath` or `npyrandom` libraries. If the dependencies are installed only for the build arch, the interpreter will fail entirely and meson will never even configure the build.

## Resolution attempts

I have tried a couple simple workarounds, to no avail:
1. Install only the dependencies only for the build arch. On Void Linux, we [set a lot of environment variables](https://github.com/void-linux/void-packages/blob/master/common/build-style/python3-module.sh#L16-L20) to tell the host Python to use the sysconfig data for the build arch and also add the build root to `PYTHONPATH`, allowing the host Python to find these modules and grabbing relevant information (field sizes, shlib suffixes, etc.) for the build arch rather than the host. This *would* work with the existing meson detection for `pybind11` and `pythran`, but does not work for `numpy` because `import numpy` triggers a bunch of shared object loads and the build arch libraries are incompatible with the host. (A more targeted import, such as `from numpy.__config__ import get_include`, might work in this situation, but I haven't bothered to try.)
2. Install dependencies for host and build arches, use the detection to find paths to the host, but then manually prepend the build root to the returned paths. This *almost* works on Void Linux because we install packages for the build arch under a `/usr/<triple>` prefix that otherwise mirrors the native layout. In generally, most of the compilation commands include the right `-I` flags to find headers for the build arch. However, some commands still include paths to the host interpreter, and the [`find_library` calls](https://github.com/scipy/scipy/blob/main/scipy/meson.build#L54-L59) to identify `npymath` and `npyrandom` somehow still pick up the host versions and trigger a linker failure. I don't know enough about how meson sets the Python environment [when searching for it](https://github.com/scipy/scipy/blob/main/meson.build#L63-L65) to understand how these host paths are creeping in or why `find_library` still seems to prefer the host paths even though I add the correct paths to the search paths in that function. (I wouldn't really expect `find_library` to dig into the numpy tree to find the libraries, so it seems an additional search path is creeping in before I add one explicitly.)

## Possible fixes

Although I'm speculating, it seems a few approaches could be taken to resolve this issue, in order of decreasing "niceness":
1. Convince `numpy` (and, probably, `pybind11`) to ship pkg-config files. This is probably desirable, was mentioned in the [related meson issue], and sidesteps a lot of problems. The trouble with `pybind11` is that it wants to be entirely self-contained within the Python package tree; however, even if it ships a `.pc` file within its package tree instead of in a system-specific path, Void can probably work around the issue with relative ease. (Void already wraps pkg-config for cross builds so that it loads descriptors for the build arch and manipulates the paths appropriately.) The determination for `pythran` should just be dropped altogether (also, rather than invoking a Python interpreter to read `SCIPY_USE_PYTHRAN` from the environment, Pythran should just be a meson build option). The trouble here is backwards compatibility; if SciPy will build with old versions of NumPy or pybind11, it will still need fallback detection. Hence...
2. Existing logic can be improved, even if only as a fallback for older versions of dependencies. I'm not sure what this should look like, but reading variables from the environment might make sense (*e.g.*, `NUMPY_ROOT`, `PYBIND11_INCLUDE_DIR`); when these variables are defined, they are used as-is; otherwise, the existing interpreter invocations can provide sensible defaults for native builds. Of course, this assumes that `find_library` can be made to find NumPy libraries for the build arch even though it now prefers the host versions.
3. The [existing search for a Python interpreter](https://github.com/scipy/scipy/blob/main/meson.build#L64) could allow a custom path rather than always using the default, which seems to be the same interpreter that is running meson. This might allow some clever wrapping of the interpreter but is probably an incomplete (and maybe completely ineffective) solution. For example, no amount of sensible wrapping will allow `print(numpy.get_include())` to dump some modified path; however, a successful `from numpy.__config__ import get_include` using the build-arch NumPy might make this workable if ugly.

## Related issue

This issue was opened in response to https://github.com/mesonbuild/meson/issues/9598#issuecomment-1201475862 as a means to track SciPy specifics and provide a link target for inclusion in https://github.com/scipy/scipy/issues/14812.

cc: @rgommers

**Comment by rgommers** [comment]
@ahesford thanks a lot for the detailed bug report. We are just working through resolving the same issue for Buildroot in gh-14812. Still like to commit a temporary fix there, but at least there was success when carrying one patch with a hardcoded path for the two static libraries coming from NumPy. This is a cleaner and more dedicated issue, so I may continue here.

>  but reading variables from the environment might make sense (_e.g._, `NUMPY_ROOT`, `PYBIND11_INCLUDE_DIR`); when these variables are defined, they are used as-is; otherwise, the existing interpreter invocations can provide sensible defaults for native builds.

That seems like a useful thing to do as a workaround at least.

> 3\. For example, no amount of sensible wrapping will allow `print(numpy.get_include())` to dump some modified path; however, a successful `from numpy.__config__ import get_include` using the build-arch NumPy might make this workable if ugly.

I'm going to really try to not need to run the target Python interpreter for the final solution.

> 1. Convince `numpy` (and, probably, `pybind11`) to ship pkg-config files. This is probably desirable, was mentioned in the [related meson issue], and sidesteps a lot of problems.

Yes, I do want to do this for `numpy`.

**Comment by rgommers** [comment]
> as a means to track SciPy specifics and provide a link target for inclusion in #14812.

I added Void Linux to the issue description there.

> The trouble here is backwards compatibility; if SciPy will build with old versions of NumPy or pybind11, it will still need fallback detection. Hence...

My goal is for that fallback detection to live inside Meson; `scipy/meson.build` should only contain `dependency('numpy')` and similar look-ups. That will work with older numpy versions too.

**Comment by eli-schwartz** [comment]
> 1. Convince `numpy` (and, probably, `pybind11`) to ship pkg-config files. This is probably desirable, was mentioned in the [related meson issue], and sidesteps a lot of problems. The trouble with `pybind11` is that it wants to be entirely self-contained within the Python package tree; however, even if it ships a `.pc` file within its package tree instead of in a system-specific path, Void can probably work around the issue with relative ease. (Void already wraps pkg-config for cross builds so that it loads descriptors for the build arch and manipulates the paths appropriately.)

By sheer coincidence, I had already submitted https://github.com/pybind/pybind11/pull/4077

Note that pybind11 already installs cmake files, and Meson can pick those up, although again it installs them to the self-contained python package tree, which is a bit painful. There is also a Meson wrap, so `dependency('pybind11')` will find something via cmake, if you manually set some cmake variables, and otherwise download a private copy. I'm generally hoping to make this better (particularly, you can install pybind11 via `cmake ...... && make install`, which installs to the system instead and which I believe is vastly superior).

**Comment by ahesford** [comment]
Re: pybind11, it was only recently that they moved everything into the package tree; earlier releases installed in system locations. I can see why they might want to keep things self contained, but having the headers installed in a system location *does* make things a bit easier from our perspective.

**Comment by rgommers** [comment]
Randomly I just ran into a pkg-config related file for `libnpymath` that actually does get installed. Doesn't help much, because there's no corresponding one for `libnpyrandom.a`:

```ini
$ cat numpy/core/lib/npy-pkg-config/npymath.ini 
[meta]
Name=npymath
Description=Portable, core math library implementing C99 standard
Version=0.1

[variables]
pkgname=numpy.core
pkgdir=/home/rgommers/code/tmp/buildroot-scipy/oscimp_br2_external/buildroot/output/host/aarch64-buildroot-linux-gnu/sysroot/usr/lib/python3.10//site-packages/numpy/core
prefix=${pkgdir}
libdir=${prefix}/lib
includedir=${prefix}/include

[default]
Libs=-L${libdir} -lnpymath
Cflags=-I${includedir}
Requires=mlib

[msvc]
Libs=/LIBPATH:${libdir} npymath.lib
Cflags=/INCLUDE:${includedir}
Requires=mlib
```

**Comment by henryiii** [comment]
> Re: pybind11, it was only recently that they moved everything into the package tree; earlier releases installed in system locations. 

Both options are available. `pybind11-global` installs to the system locations (intended for use in pyproject.toml). You can get both via `pybind11[global]`.

**Comment by rgommers** [comment]
I tried to get rid of `libnpymath.a` and `libnpyrandom.a` completely (see https://github.com/numpy/numpy/issues/20880#issuecomment-1206846026), but that's too complex unfortunately - at least on the timescale in which we want to get this fixed in scipy `1.9.1`.

**Comment by rgommers** [comment]
> I tried to get rid of `libnpymath.a` and `libnpyrandom.a` completely (see [numpy/numpy#20880 (comment)](https://github.com/numpy/numpy/issues/20880#issuecomment-1206846026)), but that's too complex unfortunately - at least on the timescale in which we want to get this fixed in scipy `1.9.1`.

For the record: I'm actually giving this a serious go, because it would be much healthier for both NumPy and SciPy to actually get this done. The changes are just too large to be backported to `1.9.x`, so we need a workaround there.

**Comment by ahesford** [comment]
I've really only looked at `libnpymath` for two minutes, but it leaves me scratching my head. It seems like a bunch of convenience wrappers to paper over implementation differences in low-level math functions that should be provided by a C runtime. While NumPy might need this compatibility layer for itself, I fail to understand why this should be exposed as a public library. The purpose of the NumPy C API is to allow compiled Python extensions (like SciPy) to interact with NumPy data. Let those extensions worry implementation disparities for low-level functions on their own.

In short, I agree with the argument in numpy/numpy#20880 that `libnpymath` should be stripped *way* back and become nothing more than a convenience header easily vendored by SciPy.

**Comment by rgommers** [comment]
> I fail to understand why this should be exposed as a public library.

A matter of history I'd say. Pre C99 one needed all that code, and C99 support in MSVC is quite recent. And that code requires a good amount of build system support, which comes from `numpy.distutils` and would be very difficult to reimplement (at least before we had Meson).

**Comment by charris** [comment]
There was also a problem with Windows compilers, they didn't always provide functions that could be passed into the ufuncs, but rather used inlined instructions.  Hence the renaming with the `npy_*` prefix. That was back around the first NumPy releases. It was a while before we could get things working on Windows.

**Comment by doronbehar** [comment]
Now, since https://github.com/pybind/pybind11/pull/4077 has landed in it's latest release. I was thinking of adding support for finding the include path via the pkg-config path, but I noticed that it doesn't really contain the prefix of the python package. And. as mentioned here earlier, it's not installed in a typical pkg-config path - meaning that in order to make `pkg-config` find it, a user would need to define `PKG_CONFIG_PATH` appropriately.

Would you be willing to accept a simple PR that will add support for environmental variables as default include paths? As suggested earlier?

**Comment by henryiii** [comment]
> it doesn't really contain the prefix of the python package

What do you mean? Is it something that we can fix? The only requirement is that it is inside site-package's pybind11 folder in the normal package (it can be anywhere in the global package).

**Comment by doronbehar** [comment]
> > it doesn't really contain the prefix of the python package
> 
> What do you mean?

I'll admit that I haven't noticed and/or tested the global installation option when writing my previous comment. In anycase, for the none global installation, the `<prefix>/lib/python3.10/site-packages/pybind11/share/pkgconfig/pybind11.pc` file has:

```
prefix=${pcfiledir}/../../
includedir=${prefix}/include

Name: pybind11
Description: Seamless operability between C++11 and Python
Version: 2.10.1
Cflags: -I${includedir}
```

Which is nice but not a huge improvement because this `pkgconfig` directory is not located in a standard location - meaning that even if scipy would have read the `includedir` [pkg config variable](https://mesonbuild.com/Reference-manual_returned_dep.html#depget_pkgconfig_variable), a user trying to cross compile scipy with a non global pybind11 installation will need to add the correct path to `PKG_CONFIG_PATH` when running `meson setup`.

However, I tested the global installation with Nix and I encountered issues which perhaps it'll be better to outline in pybind's issue tracker. I will say the the global installation did generated a useful `pybind11.pc` file with absolute values to the `prefix` and `includedir`.

So to summarize, the new `pybind11-global` can be useful for scipy, but it still needs some work. I'd still like to get feedback for the idea of simply reading plain environment variables here in `scipy`'s meson files.

**Comment by doronbehar** [comment]
All of my comments about `pybind11` were a mess. Now I noticed that running `make install` when building `pybind11` (not `pybind11-global`), does create a correct `pybind11.pc` file at a standard location, which is great for Nix! Sorry for the noise..

**Comment by eli-schwartz** [comment]
My pybind11 PR was about adding pkg-config ***at all***.

It included support for relocation because that's what the pybind11 team wants, and I agreed it's fair enough to want it -- but for distro packaging I'd strongly recommend installing via cmake, not setup.py (or else do both) and use system-integrated pkg-config files (and cmake config files as well :p) so that this all ends up in the system location.

At some point I'm probably going to add a custom dependency lookup method for pybind11 to Meson -- it will try for system pkg-config or cmake, but also handle pybind11-config, and that will allow SciPy to do `dependency('pybind11')`, which will look for system versions installed via e.g. NixOS, and then if those cannot be found it will also happily use a virtualenv-installed pyproject.toml build dependency that injects pybind11-config into the build `$PATH`.

This should then work nicely in a wide variety of circumstances, including cross compilation. In fact it already mostly does that, and the exception is "detect it via pip install in a virtualenv", so I'm waiting until I can offer a SciPy patch that both improves support and simplifies the code. ;)

**Comment by rgommers** [comment]
> The only requirement is that it is inside site-package's pybind11 folder in the normal package

@henryiii I looked into this before, concluded that it's kinda pointless to install pkg-config files in site-packages (I'd like to install them for NumPy), and that it should in principle be possible to install them to the correct location under `<prefix>/include/` by using https://peps.python.org/pep-0427/#the-data-directory. Does that seem feasible to you?

**Comment by eli-schwartz** [comment]
@rgommers,

I don't disagree with you, but pybind11 already has `pybind11_global` for the exact same situation except replacing the word "pkg-config" with "cmake" everwhere.

It installs a second copy of all data files, IIRC, and then depends on `pybind11` for the -config script and importable module. That's kinda extraneous but relatively speaking not by all *that* much.

**Comment by henryiii** [comment]
No, `pip install package` should not install things outside of site packages. If a user does this without being inside a virtual environment, it messes with the base system. Try this in brew then type "brew doctor", and it will complain about mysterious files that it does not control showing up in the root. This is why pybind11 has a `pybind11-global` copy - the normal pybind11 doesn't touch anything outside of site-packages. The "global" exists entirely for this purpose - it is allowed to put things outside of site-packages.

There are two better options, IMO, though. One is to use an entry-point for pkg-config. Tools like meson-python and scikit-build-core could check for this entrypoint, and add the folders specified. I briefly had this for cmake files, but removed it while we design a proper name and system. But it did work, there's a PR for pybind11 and history in scikit-build-core where it was working.

Second option requires pkg-config's search to be smart enough to detect `<package>/...` as a pkg-config file - CMake does this (`<package>/share/...`), so if you add site-packages to the correct CMake variable, it can detect pybind11's config. This is a bit more limited (it has have the same name as the package it's in), but works in CMake's case, and is how scikit-build-core finds pybind11 currently. Not sure if pkg-config has the same search option.

**Comment by henryiii** [comment]
>  exact same situation except replacing the word "pkg-config" with "cmake" everwhere

Is the pkg-config not in that one too? It should be, IMO.

**Comment by eli-schwartz** [comment]
Sorry, I meant "the original rationale for `pybind11_global` was this, except `s/pkg-config/cmake/` since cmake was all there was at the time, so when I implemented pkg-config I just treated it the same as cmake".

pybind11_global does indeed have pkg-config files in the data directory, and that's part of the tests we hooked up. :)

**Comment by rgommers** [comment]
> but pybind11 already has `pybind11_global` for the exact same situation

Ah, I didn't realize that that's an actual [package on PyPI](https://pypi.org/project/pybind11-global/). Thanks for elaborating.

> No, `pip install package` should not install things outside of site packages. If a user does this without being inside a virtual environment, it messes with the base system.

That's explicitly the point of the `.data` dir in PEP 427 though. So is that one of those "good idea in principle, but you must not use this, and no we won't bother to update the PEP" things?

**Comment by eli-schwartz** [comment]
I think different people have different ideas of what it should mean, probably.

For example, I would consider the data directory an officially sanctioned method for messing with the base system on the grounds that it is my desire to mess with the base system.

Obviously, `pip install package` installs things outside of site-packages and people are okay with this, because `console_scripts` is installed outside of site-packages. It's just installed to the "scripts" directory instead of the "data" directory, so it's arbitrarily declared to be acceptable to install to the scripts directory specifically. After all, scripts are really common. Common enough that I suppose `brew doctor` doesn't consider it something to complain about.

**Comment by henryiii** [comment]
I wouldn't say there aren't places where you can use it (`pybind11_global` does use it!), but it's somewhat discouraged due to the ability to do exactly what it was supposed to do - install to any arbitrary place at the root of your python environment - which includes `/` if you are not in a venv. That's why we have a "safe" `pybind11` package and a "you know what you are doing" `pybind11_global` package.
